### PR TITLE
Checking sandbox for ArrayAccess

### LINKED ISSUE #106: Sandbox array bug & fix
# Situation: 

An unexpected behavior regarding arrays, in Twig_Template::getAttribute() if a referenced object is an array then security checking is skipped. It is allright, but in addition to checking with is_array() Twig also checks for objects that implement ArrayAccess and if so then is considered an array and security checking is again skipped. 
# Issue:

It doesn't seem a good behavior, because there are a lot of objects that implement ArrayAccess and a security policy must be enforced on them. For example I found about this behavior while feeding the template with some active record objects and the base AR class implements ArrayAccess to allow access for model attributes using array like syntax. But still the security policy needs to be enforced especially since AR objects are tightly linked to the database and I don't want users to have any ideas... 
# Resolution:

I think the best way is to differentiate between native arrays (check with is_array()) and objects that implement ArrayAccess. If it is later, then security policy must be enforced.

### PR DESCRIPTION
**Pull request by CarsonF**
Fixes #106

Allows this to be prevented in sandbox:

``` twig
{{ app.secret_key }} {# app is \Pimple #}
```

### CODE REVIEW

**Comment by stof** [comment]
@jpauli can you review the changes in the C extension ?

**Comment by CarsonF** [comment]
Thinking about it more I'm realizing that this doesn't allow for distinction on which array item. i.e., All `Pimple` services/parameters can be allowed or none.

I'm thinking this:
```php
->checkMethodAllowed($object, 'offsetGet');
```
should be replaced with this:
```php
->checkPropertyAllowed($object, $arrayItem);
```

@stof thoughts?

**Comment by stof** [comment]
@fabpot what do you think about this ?

File: test/Twig/Tests/EnvironmentTest.php [hunk 1]
```diff
@@ -180,13 +180,12 @@ public function testExtensionsAreNotInitializedWhenRenderingACompiledTemplate()
 
     public function testAutoReloadCacheMiss()
     {
-        $template_name = __FUNCTION__;
-        $template_content = __FUNCTION__;
+        $templateName = __FUNCTION__;
+        $templateContent = __FUNCTION__;
```
Comment (OskarStark) [inline test/Twig/Tests/EnvironmentTest.php]: renaming those variables should be done in a separate PR, it just increases the diff ... :(

Comment (OskarStark) [inline test/Twig/Tests/EnvironmentTest.php]: ah sorry, maybe those lines come from the commit of @fabpot , which is in this PR?

Comment (CarsonF) [inline test/Twig/Tests/EnvironmentTest.php]: I didn't rename them, @fabpot did. I'll rebase

Comment (OskarStark) [inline test/Twig/Tests/EnvironmentTest.php]: :+1:  thank you @CarsonF

**Comment by CarsonF** [comment]
Silex is EoL and container shouldn't be used in Twig anyways.
